feat(web): Start of Sentence support - part 1 🕊

mcdurdin opened the pull request:
Relates to #3621.

* Adds basic support for `begin newContext` and `begin postKeystroke` to compiler, including `gn` and `gpk` entry points.
* Adds basic support for `begin newContext` to KeymanWeb, calling into `gn` entry point, for mobile, touch and embedded contexts.

TODO in this PR:

* Compiler:
   - [x] `readonly` semantics
   - [x] usage constraints
   - [x] `&layerChanged` system store tests
* KeymanWeb:
   - [x] `begin postKeystroke` support'
   - [x] verify that `postKeystroke` fires after model events (#6004)
   - [ ] banner should be updated after newcontext, not before. (note: this may be an unrelated issue, as the banner does not currently respect layer selection for its hints.)

For follow-up pull requests:

* Developer IDE
   - [x] Support `newContext`, `postKeystroke` and `&layerChanged` tokens (#5984)
* Help
   - [x] Documentation for `newContext`, `postKeystroke` and `&layerChanged` (keymanapp/help.keyman.com#492)
   - [x] Full guide for Caps Lock support (keymanapp/help.keyman.com#492)
* Caps Lock Layer support (#3620)
   - [x] Caps Lock Layer (#5988, #5989)

* Keyman Core
   - [ ] `&layerChanged` always `0`

Optional for v15:

* KeymanWeb
   - [ ] Automated tests for start-of-sentence
* Keyman Core
   - [ ] `begin newContext`, `begin postKeystroke` support

# User Testing

### **SUITE_BASELINE:** Ensure nothing is broken!

* **TEST_BASELINE:** Run through a set of basic keyboard usage tests with SIL Eurolatin, Khmer Angkor, and other keyboards of your choice, to ensure that existing functionality has not changed. Focus on the touch keyboard interactions. Do this on Android and iOS (simulators okay), and KeymanWeb (unminified test page).

### **SUITE_START_OF_SENTENCE:** Test start-of-sentence detection

This PR includes a simple English keyboard that should detect start of sentence, called start_of_sentence_3621.kmn. 

[start_of_sentence_3621.kmp](https://drive.google.com/uc?id=1WbiesUOgI5Nby8xJM_vSy-Ip_-98nLCf&export=download)

  - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).
  - It should drop back to the base layer after typing a single capital letter.
  - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.
  - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).
  - It should continue to allow the user to switch layers at any time.
  - It should not switch layers automatically apart from in these cases.

Test that the keyboard behaves as expected in text fields, when you move the cursor, when you switch between fields with and without text, and as you type.

* **TEST_SOS_IOS:** Run the above tests in Keyman for iOS (simulator okay) - system keyboard and in-app
* **TEST_SOS_ANDROID:** Run the above tests in Keyman for Android (virtual okay) - system keyboard and in-app
* **TEST_SOS_WEB:** Run the above tests in KeymanWeb, Prediction UI testing page - on any touch platform (emulation in Chrome okay). Check that there are no script errors.

darcywong00 commented on common/core/web/input-processor/src/text/inputProcessor.ts:
```diff
@@ -237,6 +258,20 @@ namespace com.keyman.text {
+        // or by the touch layout 'nextlayer' control.
+        this.keyboardProcessor.layerChangedStore.set(startingLayerId == this.keyboardProcessor.layerId ? '0' : '1');
+
+        let postRuleBehavior = this.keyboardProcessor.processPostKeystroke(keyEvent.device, outputTarget);
```
Is `keyEvent.device` different from `this.device`? (l. 64)

mcdurdin replied:
I don't think it is different, just another way of accessing the device.

darcywong00 commented on common/core/web/input-processor/src/text/inputProcessor.ts:
```diff
@@ -50,11 +52,28 @@ namespace com.keyman.text {
+     processNewContextEvent(outputTarget: OutputTarget): RuleBehavior {
+      const ruleBehavior = this.keyboardProcessor.processNewContextEvent(this.device, outputTarget);
+
+      if(ruleBehavior != null) {
```
```suggestion
      if(ruleBehavior) {
```

darcywong00 commented on windows/src/test/unit-tests/kmcomp/test.bat:
```diff
@@ -52,25 +52,33 @@ call :should-pass "#2241: &CasedKeys (chars)" test_casedkeys_chars.kmn || goto :
 call :should-fail "#2241: &CasedKeys (mnemonic 1)" test_casedkeys_mnemonic_1.kmn || goto :eof
 call :should-fail "#2241: &CasedKeys (mnemonic 2)" test_casedkeys_mnemonic_2.kmn || goto :eof
-call :should-pass "#2241: &CasedKeys (mnemonic 3)" test_casedkeys_mnemonic_3.kmn || goto :eof
+call :should-fail "#2241: &CasedKeys (mnemonic 3)" test_casedkeys_mnemonic_3.kmn || goto :eof
```
Is this test supposed to have been always failing?

mcdurdin replied:
Yes, I believe so. I think that the should-pass test was actually FAILing but if you look at (line 77)[https://github.com/keymanapp/keyman/pull/5963/files/83eb8395d69683ac33ce878067556d8e9706bbd1#diff-27713d4507c7fb0f20b201e10ea21858b08f16e2abee737ae3490f18d405bad6R77] you can see it was masking the failure. Ooops.

darcywong00 commented on windows/src/developer/kmdecomp/savekeyboard.cpp:
```diff
@@ -199,13 +204,14 @@ PWCHAR ExtString(PWCHAR str)
 
 	bufpointer = !bufpointer;
 	p = buf[bufpointer]; *p = 0;
+  PWCHAR q = p;
```
Some of these look like changes from #5970

mcdurdin replied:
Yes, apologies, I merged that in early. You can exclude commits 2de8c145c076b361494599fa1fd831809350c7c2 and eb2892ec01091b5c3e77b11397c170ec42ede098 from your review.

darcywong00 commented on common/core/web/keyboard-processor/src/text/kbdInterface.ts:
```diff
@@ -948,20 +950,59 @@ namespace com.keyman.text {
+      return this.process(this.activeKeyboard.process.bind(this.activeKeyboard), outputTarget, keystroke);
+    }
+
+    private process(callee, outputTarget: OutputTarget, keystroke: KeyEvent): RuleBehavior {
```
Is `callee` the same as `arguments.callee`? I see the latter has some limitations
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments/callee

mcdurdin replied:
No, `callee` is not related to `arguments.callee`. They just happen to have the same name... `callee` here is the function that will be called, see L1028.

darcywong00 commented on common/core/web/keyboard-processor/src/text/kbdInterface.ts:
```diff
@@ -948,20 +950,59 @@ namespace com.keyman.text {
+     * @param       {Object}        outputTarget   The target receiving input
+     * @param       {Object}        keystroke      The input keystroke (with its properties) to be mapped by the keyboard.
+     * Description  Calls the keyboard's `begin newContext` group
+     * @returns     {number}        0 if no match is made, otherwise 1.
```
Do these 3 processX() methods return `RuleBehavior`?

Or is `this.process()` returning `0` or `1`?

mcdurdin replied:
Hmm, yes, the old comment on line 957(989) was wrong!

darcywong00 commented on common/test/keyboards/start_of_sentence_3621/source/start_of_sentence_3621.kmn:
```diff
@@ -0,0 +1,50 @@
+﻿c This keyboard provides a reasonably complete example of
+c start-of-sentence detection, for use in testing #3621.
+store(&VERSION) '10.0'
```
Should this require `15.0` now?

mcdurdin replied:
No, I think it should gracefully degrade because there are no breaking file format changes.

darcywong00 commented:
Is the TODO for KeymanWeb on this PR resolved?
>  verify that postKeystroke fires after model events

mcdurdin commented:
> Is the TODO for KeymanWeb on this PR resolved?
> 
> > verify that postKeystroke fires after model events

I just tested this, and it doesn't currently fire. Reviewing.

darcywong00 commented on web/source/dom/domEventHandlers.ts:
```diff
@@ -361,8 +363,19 @@ namespace com.keyman.dom {
+        // at least on Chrome
+        window.setTimeout(() => {
+          //console.log('processNewContextEvent called from focus');
+          com.keyman.singleton.core.processNewContextEvent(outputTarget);
```
Is this the same as? (`keyman` from L333)
```suggestion
          keyman.core.processNewContextEvent(outputTarget);
```

darcywong00 replied:
ah, maybe it's not in the scope of setTimeout()

mcdurdin replied:
It should be in scope and should work.

darcywong00 reviewed (approved):
lgtm

MakaraSok commented:
## SUITE_BASELINE: Ensure nothing is broken!
- **TEST_BASELINE: FAILED** Everything seems to work well in vertical mode, but not in horizontal mode.
   - iOS: it's working OK in vertical mode. In horizontal mode, however, the subkey functionality doesn't work as expected. Press and hold on a base key and then roll your finger to a character to see that it doesn't select where the finger tip is touching. 

   https://user-images.githubusercontent.com/28331388/144982969-c99978fb-66c0-43e5-8336-17b8ab4d44f5.mov
   
   - Android: the keyboard works OK right off in both vertical and horizontal modes, but then it misbehaves. It seems like the key selection doesn't focus on the expected key.

   https://user-images.githubusercontent.com/28331388/144983051-0f3c0a9c-df81-41f7-a7a0-bd0d63cf88b7.mov

## SUITE_START_OF_SENTENCE: Test start-of-sentence detection
- **TEST_SOS_IOS: FAILED** The keyboard doesn't start on a Shift layer in a new field nor after the punctuations listed above.

   <img width="70" alt="image" src="https://user-images.githubusercontent.com/28331388/144996391-8744d060-b757-4bf6-874e-c732e5eb671a.png">

- **TEST_SOS_ANDROID: FAILED** The keyboard doesn't start on a Shift layer in a new field nor after the punctuations listed above.

   <img width="70" alt="image" src="https://user-images.githubusercontent.com/28331388/144996544-b2536536-89af-4213-92eb-25514994ef75.png">

- **TEST_SOS_WEB: BLOCKED** It looks like the builds are not current as all tests are failed. @mcdurdin ptas

mcdurdin commented:
@keymanapp-test-bot SUITE_START_OF_SENTENCE TEST_SOS_IOS TEST_SOS_ANDROID TEST_SOS_WEB retest I've updated the test spec to include a link to the .kmp for these tests.

mcdurdin commented:
> * iOS: it's working OK in vertical mode. In horizontal mode, however, the subkey functionality doesn't work as expected. Press and hold on a base key and then roll your finger to a character to see that it doesn't select where the finger tip is touching.
>
> * Android: the keyboard works OK right off in both vertical and horizontal modes, but then it misbehaves. It seems like the key selection doesn't focus on the expected key.

Do you know if this happens also with the current master build (15.0.163) or is this new to this PR?

MakaraSok commented:
> > * iOS: it's working OK in vertical mode. In horizontal mode, however, the subkey functionality doesn't work as expected. Press and hold on a base key and then roll your finger to a character to see that it doesn't select where the finger tip is touching.
> > * Android: the keyboard works OK right off in both vertical and horizontal modes, but then it misbehaves. It seems like the key selection doesn't focus on the expected key.
> 
> Do you know if this happens also with the current master build (15.0.163) or is this new to this PR?

I've seen this before and it was considered as a non-issue. 
cf. https://github.com/keymanapp/keyman/issues/5964

mcdurdin commented:
> I've seen this before and it was considered as a non-issue.

Given that this particular issue occurs even in 14.0, let's not fail the start of sentence support because of this but track that separately?

SUITE_BASELINE
* TEST_BASELINE OPEN _please retest_: if there are any other issues, please report back

MakaraSok commented:
## SUITE_START_OF_SENTENCE
- **TEST_SOS_IOS: FAILED**  
  * in-app:
     - It should set the keyboard to the Shift layer at the start of a text field, after full stop (.), exclamation (!) or question mark (?) and one or spaces (e.g. . ).
     > _OK_
     - It should drop back to the base layer after typing a single capital letter.
     > _OK_
     - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.
     > _Not working. The keyboard switches to the default layer after a digit is pressed._

     https://user-images.githubusercontent.com/28331388/145164915-8c310349-bbd4-491d-8270-da44591886e6.mov
     
     - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

     > _Not working._
     <img width="139" alt="image" src="https://user-images.githubusercontent.com/28331388/145172792-6de7d44d-9a41-44e3-8405-2b9534a850d1.png">
     
     - It should continue to allow the user to switch layers at any time.

     > _OK_
     
     - It should not switch layers automatically apart from in these cases.

     > _OK_
     
  * system:
     - It should set the keyboard to the Shift layer at the start of a text field, after full stop (.), exclamation (!) or question mark (?) and one or spaces (e.g. . ).

     > _Not working. (See the screen recording below)._
     
     https://user-images.githubusercontent.com/28331388/145175641-d803e500-0ec7-49db-8223-add0fb35bd21.mov
     
     - It should drop back to the base layer after typing a single capital letter.

     > _Not working. See the screen recording right above._
     
     - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

     > _Not working. Neither would switch the keyboard to the default layer._

     https://user-images.githubusercontent.com/28331388/145176274-c33e3526-bbb4-45cb-ab5f-60610cf69252.mov
     
     - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

     > _The keyboard stays on Shift layer always. See the first screen recording of this test set._
     
     - It should continue to allow the user to switch layers at any time.

     > _OK_
     
     - It should not switch layers automatically apart from in these cases.

     > _OK_
     
- **TEST_SOS_ANDROID: FAILED**
  * in-app:
     - It should set the keyboard to the Shift layer at the start of a text field, after full stop (.), exclamation (!) or question mark (?) and one or spaces (e.g. . ).

     > _Not working. The Shift layer stays stuck from the get go. Switch it off and it stay of._

     https://user-images.githubusercontent.com/28331388/145177968-ddd2fb48-cd20-4a00-b4e2-5bf8f4a1fa41.mov
     
     - It should drop back to the base layer after typing a single capital letter.

     > _Not working. See the screen recording right above._
     
     - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

     > _Not working. The keyboard does switch to the default layer after typing a symbol. It stays there until manually switched._
     
     - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

     > _Not working. It's currently stuck in the currently active layer._
     
     - It should continue to allow the user to switch layers at any time.

     > _OK_
     
     - It should not switch layers automatically apart from in these cases.

     > _OK_
     
  * system:
     - It should set the keyboard to the Shift layer at the start of a text field, after full stop (.), exclamation (!) or question mark (?) and one or spaces (e.g. . ).

     > _Not working. The keyboard stays on Shift layer._

     https://user-images.githubusercontent.com/28331388/145178904-cb9d724b-afd6-4c64-ac6c-755b990908af.mov
     
     - It should drop back to the base layer after typing a single capital letter.

     > _Not working. See the screen recording right above._
     
     - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

     > _Not working. The keyboard stays on the numeric layer even after pressing a symbol key._

     - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

     > _Not working. The keyboard would stay in a chosen layer._
     
     - It should continue to allow the user to switch layers at any time.

     > _OK_
     
     - It should not switch layers automatically apart from in these cases.

     > _OK_
     
- **TEST_SOS_WEB: BLOCKED** Please provide instructions on how to test this using the keyboard package above. Thanks.

MakaraSok commented:
## SUITE_BASELINE: Ensure nothing is broken!
- **TEST_BASELINE: PASSED** no significant issue experienced. The only issue is the different rendering of shift layers on iOS and Android. iOS rendered the layer nicely, unlike Android. See below.
<img width="740" alt="image" src="https://user-images.githubusercontent.com/28331388/145323492-69a73ac9-5ac2-4296-ab40-dc95872daabb.png">

mcdurdin commented:
SUITE_START_OF_SENTENCE
* **TEST_SOS_WEB: OPEN** I've updated the instructions and fixed a link in the KeymanWeb test page; please retest :grin:

mcdurdin commented:
I'm very sorry 😭, I uploaded an old version of the start_of_sentence_3621.kmp package. I've updated the package, and tested it myself with good outcomes, so please retest with the new package.

## SUITE_START_OF_SENTENCE
* **TEST_SOS_IOS: OPEN** _please retest_
* **TEST_SOS_ANDROID: OPEN** _please retest_

MakaraSok commented:
## SUITE_START_OF_SENTENCE: Test start-of-sentence detection
- **TEST_SOS_IOS: FAILED** _Tested with iOS 15.0. mostly_
  * In-app
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > Not working! In iOS 15.0, there are two tabs audibly heard right after opening the Keyman app before the "Get Started" window appears. Click on "Done" to close the window to see the default layer of the keyboard instead of the shift layer. See the screenshot below.

      <img width="70" src="https://user-images.githubusercontent.com/28331388/145749735-95b649bf-0411-4a66-bcff-dcdea2b0e5e3.png">

      The shift layer appears only after tapping in the text area.

      <img width="70" src="https://user-images.githubusercontent.com/28331388/145749961-f5424b7e-1797-461b-b600-f3ea289d9613.png">

      Here is a screen recording for a better visual.
      
      https://user-images.githubusercontent.com/28331388/145767663-35ebaf40-76bf-4782-9067-092b6a45bb23.mov

      In iOS 13.7, still hear the two taps, but no "Get Started" page is shown. The keyboard layout does show the shift layer right off.

      <img width="70" src="https://user-images.githubusercontent.com/28331388/145750446-1ba50064-2613-4215-b95a-e86277436ca2.png">
       
    - It should drop back to the base layer after typing a single capital letter.
    
      > Working!

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Not working! Pressing any key on the numeric layer moves the keyboard back to the default layer.
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Not working! The keyboard doesn't stay on shift layer after typing two capital letters in a row.

      <img width="70" src="https://user-images.githubusercontent.com/28331388/145758453-99d793a5-add4-4a0e-aac3-51e6c7a469dc.png">

    - It should continue to allow the user to switch layers at any time.

      > Working!
    
    - It should not switch layers automatically apart from in these cases.

      > Working!

    _Note that the default layer is shown when the Keyman app is opened from standby. It is expected to have shift layer shown instead._
    
    https://user-images.githubusercontent.com/28331388/145761077-d2fc57ed-2912-4244-8d8c-9825a70cad33.mov
      
  * As a system keyboard:
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > Working. 
       
    - It should drop back to the base layer after typing a single capital letter.

      > Working.

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Working.
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Working.

    - It should continue to allow the user to switch layers at any time.

      > Working.

    - It should not switch layers automatically apart from in these cases.

      > OK.

    _After a line break (pressing Enter key), the shift layer should be shown, but it still stays in the default layer. Please consider adding this as a new functionality._

MakaraSok commented:
## SUITE_START_OF_SENTENCE: Test start-of-sentence detection
      
- **TEST_SOS_ANDROID: FAILED**
  * In-app:
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > The shift layer does show up after closing the "Get Started" page, but the longpress keys are not working. The subkeys are not popped up.

      https://user-images.githubusercontent.com/28331388/145778257-095a7f54-875a-4bb6-ad40-516414def220.mov

    - It should drop back to the base layer after typing a single capital letter.

      > Working as expected.

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Working as expected. Note also that the longpress keys on this layer are not working. Subkeys don't pop up!
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Working as expected.

    - It should continue to allow the user to switch layers at any time.

      > Working as expected.

    - It should not switch layers automatically apart from in these cases.

      > OK

  * As a system keyboard:
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > The shift layer does show up after closing the "Get Started" page, but the longpress keys are not working. The subkeys are not popped up.

      https://user-images.githubusercontent.com/28331388/145779470-a428b65f-11e2-49a7-b02b-bd5b2b80d1c0.mov
       
    - It should drop back to the base layer after typing a single capital letter.

      > Working as expected.

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Working as expected.
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Working as expected.

    - It should continue to allow the user to switch layers at any time.

      > OK

    - It should not switch layers automatically apart from in these cases.

      > OK
- **TEST_SOS_WEB: FAILED?** _Tested on Windows 10, Chrome Version 96.0.4664.93 (Official Build) (64-bit)_
  - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

    > Not working. At the start of the text field, the default (unshift) layer is shown.

    Here is what I see in the console:
    <img width="300" src="https://user-images.githubusercontent.com/28331388/145746648-2507172f-e1db-4e19-a160-080ad0132d51.png">
     
  - It should drop back to the base layer after typing a single capital letter.

    > Not working.

  - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

    > Not verifiable as the is no numeric layer. In anyway, no auto-switch of keyboard layer at any point though.
    
  - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

    > Not verifiable.

  - It should continue to allow the user to switch layers at any time.

    > OK

  - It should not switch layers automatically apart from in these cases.

    > OK

mcdurdin commented:
The web test should be run with the correct keyboard, which is the start_of_sentence_3621 keyboard available in the Prediction UI test page. For the current build, that's https://build.palaso.org/repository/download/Keymanweb_TestPullRequests/300375:id/testing/prediction-ui/  The keyboard can be found under language "English...", "Start of Sentence test"

@keymanapp-test-bot retest SUITE_START_OF_SENTENCE TEST_SOS_WEB

mcdurdin commented:
> * [TEST_SOS_ANDROID] The shift layer does show up after closing the "Get Started" page, but the longpress keys are not working. The subkeys are not popped up.

@MakaraSok  I am not able to reproduce this problem with the keyboard, after testing both on my device and on the emulator. Can you check again, and if you can reproduce it, perhaps we can look more closely at the difference between your test and mine?

@keymanapp-test-bot retest SUITE_START_OF_SENTENCE TEST_SOS_ANDROID

MakaraSok commented:
## SUITE_START_OF_SENTENCE: Test start-of-sentence detection   
- **TEST_SOS_ANDROID: FAILED** Longpress not working.
  * In-app:
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > OK, but the longpress keys are not working. The subkeys are not popped up.

     https://user-images.githubusercontent.com/28331388/146702344-0caad2d4-fe8f-4245-87f7-e9c7c5ca792e.mov


    - It should drop back to the base layer after typing a single capital letter.

      > Working as expected.

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Working as expected. Note also that the longpress keys on this layer are not working. Subkeys don't pop up!
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Working as expected.

    - It should continue to allow the user to switch layers at any time.

      > Working as expected.

    - It should not switch layers automatically apart from in these cases.

      > While on numeric layer, pressing the spacebar after a digit switch the keyboard layout to default.

  * As a system keyboard:
    - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

      > The same behavior as in-app. The long press doesn't work either.

     https://user-images.githubusercontent.com/28331388/146702743-f3aed9ab-662a-4d40-ad8d-a32dbcad2217.mov

       
    - It should drop back to the base layer after typing a single capital letter.

      > Working as expected.

    - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

      > Working as expected.
      
    - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

      > Working as expected.

    - It should continue to allow the user to switch layers at any time.

      > OK

    - It should not switch layers automatically apart from in these cases.

      > It switches to the default layer while on the numeric layer after pressing the spacebar after a digit.


- **TEST_SOS_WEB: PASSED** _Tested on Windows 10, Chrome Version 96.0.4664.93 (Official Build) (64-bit)_
  - It should set the keyboard to the Shift layer at the start of a text field, after full stop (`.`), exclamation (`!`) or question mark (`?`) and one or spaces (e.g. `. `).

    > Working OK.

    I don't know what to look out for, but here is what I see:
    <img width="158" alt="image" src="https://user-images.githubusercontent.com/28331388/146701063-754adeaf-0962-41e0-8f36-c7e3fda3033f.png">

     
  - It should drop back to the base layer after typing a single capital letter.

    > Working OK.

  - It should keep the keyboard in the numeric layer after typing a digit, but move back to base layer for all other symbols.

    > Working OK.
    
  - It should keep the keyboard in the shift layer after typing two capital letters (note that this is a temporary testing behaviour until we support a CAPS layer in an upcoming PR).

    > Working OK.

  - It should continue to allow the user to switch layers at any time.

    > OK

  - It should not switch layers automatically apart from in these cases.

    > On the numeric layer, press a space after a digit switches the keyboard layer to default.

mcdurdin commented:
Tracking user testing failures in two separate issues and will go ahead and merge this and the entire 🕊 start-of-sentence chain, given I think these two issues can probably be resolved in beta:

* #6112 
* #6113

Note: I have managed to reproduce #6112, but have not reproduced #6113.

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 15.0.179-alpha](https://keyman.com/downloads/releases/alpha/15.0.179)
